Document and restructure libnetplan's public API symbols

slyon opened the pull request:
## Description
Please review commit-by-commit.

It's mostly adding Doxygen comments to our public functions.
But also dropping one public symbol from the API (`netplan_generate`), which is not properly implemented and still unused, so let's get rid of it until we grow a need for it and can have it implemented the correct way.

Finally, I'm moving around the documented declarations into their related header files:
- parser.h / parser-nm.h: Functions related to `NetplanParser*`
- state.h: Function related to `NetplanState*`
- netdef.h: Functions related to `NetplanNetDefinition*`
- utils.h: Function not related to any specific object
- types.h: data structures, no functions

#### Note
You can ignore the final commit ([abi-compat: Update for dropped 'netplan_generate' symbol](https://github.com/canonical/netplan/pull/438/commits/4bff7e6d481d0e1c73c717710abbbf5b66c1bff2)), which is auto-generated noise to `abi-compat/jammy_1.0.xml`.

FR-6079

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).
- [x] New/changed keys in YAML format are documented.
- [x] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

daniloegea commented on include/netplan.h:
```diff
@@ -79,63 +151,185 @@ netplan_state_update_yaml_hierarchy(
 netplan_netdef_has_match(const NetplanNetDefinition* netdef);
 
+/**
+ * @brief   Gheck if a @ref NetplanNetDefinition matches on given interface parameters.
```
Typo, s/Gheck/Check/

daniloegea commented on include/types.h:
```diff
@@ -62,14 +70,19 @@ typedef enum {
+/// See @ref netplan_state_get_netdef, @ref netplan_netdef_get_id and related accessor functions.
+typedef struct netplan_net_definition NetplanNetDefinition;
+
+/// Renderer backends, supported by Netplan.
```
Remove the comma?

daniloegea commented on include/types.h:
```diff
@@ -93,56 +115,51 @@ struct _NetplanStateIterator {
- * where the error was found.
+/**
+ * @brief   Errors for domain @ref NETPLAN_VALIDATION_ERROR.
+ * @details Such are expected to contain only the YAML file name
```
"Such errors"?

daniloegea commented on include/util.h:
```diff
@@ -15,11 +16,10 @@
+/**
+ * @file util.h
+ * @brief A set of generic helper functions that can be used when working with
+ *        Netplan data structures. Like handling of @ref NetplanError data.
```
s/Like/Such as/?

daniloegea reviewed (approved):
Looks good. I just left a few suggestions and notes about typos.

slyon commented:
Thanks, those were some nice finds! I fixed it all as suggested and rebased. Let's get this merged, once CI is green.

@rkratky Feel free to still leave your comments in here. If anything, I can fix it in a follow-up PR.

slyon commented:
The NetworkManager CI is failing again... But it is unrelated to these changes and needs fixing independently in the CI images.
